### ISSUE TITLE
Model parameters that themselves have a model

### ISSUE THREAD

**Issue by embray** [astropy/astropy#4013]
@wkerzendorf has raised the issue that it would be useful in some cases to take the value for a parameter from an already known distribution (based on the input) rather than assigning it directly.  You can already do this in an ad-hoc manner of course.  Something like (if I understand the use case correctly):

```python
x = np.arange(-2, 2, 0.1)
offset_dist = Gaussian1D(1, 0.0., 0.2)
model = Polynomial1D(degree=3, c0=offset_dist(x), c1=1, c2=2, c3=3)
y = model(x)
```

I thought about this when working on model composition--it is possible to pass the output of one model as the input to another model.  But it wasn't possible to pass the output of a model in as one of the parameters, at least not directly.

I think, if people think this would be useful, this could be implemented fairly simply--instead of passing in an array or scalar value for a parameter, pass in a *model* object.  Then inputs to the main model would also be passed in to any models for parameters.  

Supporting this would require changes in lots of places, but the concept (again, if I'm understanding the use case correctly) is straightforward enough.

(Also any models with a `fit_deriv` would have to be modified to support this--as long as the main model and any models for parameters have `fit_deriv` implementations though this could still derive a new `fit_deriv` from the chain rule (see also #3783)).

**Comment by mhvk** [comment]
This would be great to have! The specific use case is spectral extraction, where for each spectral bin, a gaussian is fit to get the flux in that bin. But spectra rarely align perfectly with a CCD, so the centre (and possibly width) of the gaussian would vary with spectral position.

**Comment by embray** [comment]
@mhvk Thanks for the explanation, that's good to keep in mind.

That could *almost* be done now with a `tied` constraint, but it would be tricky.  In this case the centre and width would be "tied", but to a model that depends on the spectral position, right?

**Comment by mhvk** [comment]
Right, since center and fwhm would each be tied to the output of a model, one can think of this as a (possibly complicated) tie to that model's parameters (and, as you noted, one would need to do proper error propagation with the chain rule).

**Comment by mhvk** [comment]
p.s. I like the analogy of the tie: it immediately makes clear that the "tied" parameter does not count in the degrees of freedom.

**Comment by embray** [comment]
Right--even if we don't necessarily call it a "tied" constraint in the same way those are currently implemented, it has a similar purpose and removes a degree of freedom.

**Comment by wkerzendorf** [comment]
@embray this is my attempt at making a model like this. The `evaluate` function works, but the coefficients do not show up in the model: https://github.com/wkerzendorf/extract_spectral/blob/master/extract_spectral/trace/base.py
